cloud-controller-manager/do/loadbalancers.go

```diff
@@ -439,60 +446,78 @@ func buildForwardingRules(service *v1.Service) ([]godo.ForwardingRule, error) {
+		protocol = protocolHTTP
 	}
 
-	if len(tlsPorts) > 0 {
-		if certificateID == "" && !tlsPassThrough {
-			return nil, errors.New("must set certificate id or enable tls pass through")
+	for _, port := range service.Spec.Ports {
+		// We use https for TLS, so set it explicitly if necessary.
+		if tlsPortMap[port.Port] {
+			protocol = protocolHTTPS
 		}
 
-		if certificateID != "" && tlsPassThrough {
-			return nil, errors.New("either certificate id should be set or tls pass through enabled, not both")
+		forwardingRule, err := buildForwardingRule(service, &port, protocol, certificateID, tlsPassThrough)
+		if err != nil {
+			return nil, err
 		}
+		forwardingRules = append(forwardingRules, *forwardingRule)
 	}
 
-	var forwardingRules []godo.ForwardingRule
-	for _, port := range service.Spec.Ports {
-		var forwardingRule godo.ForwardingRule
-		if port.Protocol != "TCP" {
-			return nil, fmt.Errorf("only TCP protocol is supported, got: %q", port.Protocol)
-		}
+	return forwardingRules, nil
+}
 
-		forwardingRule.EntryProtocol = protocol
-		forwardingRule.TargetProtocol = protocol
+func buildForwardingRule(service *v1.Service, port *v1.ServicePort, protocol, certificateID string, tlsPassThrough bool) (*godo.ForwardingRule, error) {
+	var forwardingRule godo.ForwardingRule
 
-		forwardingRule.EntryPort = int(port.Port)
-		forwardingRule.TargetPort = int(port.NodePort)
+	if port.Protocol != "TCP" {
+		return nil, fmt.Errorf("only TCP protocol is supported, got: %q", port.Protocol)
+	}
 
-		// TLS rules should only apply when default protocol is http or https
-		for _, tlsPort := range tlsPorts {
-			if port.Port == int32(tlsPort) {
-				forwardingRule.EntryProtocol = "https"
+	forwardingRule.EntryProtocol = protocol
+	forwardingRule.TargetProtocol = protocol
 
-				if tlsPassThrough {
-					forwardingRule.TlsPassthrough = tlsPassThrough
-					forwardingRule.TargetProtocol = "https"
-				} else {
-					forwardingRule.CertificateID = certificateID
-					forwardingRule.TargetProtocol = "http"
-				}
-				break
-			}
+	forwardingRule.EntryPort = int(port.Port)
+	forwardingRule.TargetPort = int(port.NodePort)
+
+	if protocol == protocolHTTPS {
+		err := buildTLSForwardingRule(&forwardingRule, service, port.Port, certificateID, tlsPassThrough)
+		if err != nil {
+			return nil, err
 		}
+	}
+
+	return &forwardingRule, nil
+}
 
-		forwardingRules = append(forwardingRules, forwardingRule)
+func buildTLSForwardingRule(forwardingRule *godo.ForwardingRule, service *v1.Service, port int32, certificateID string, tlsPassThrough bool) error {
+	if certificateID == "" && !tlsPassThrough {
+		return errors.New("must set certificate id or enable tls pass through")
 	}
 
-	return forwardingRules, nil
+	if certificateID != "" && tlsPassThrough {
+		return errors.New("either certificate id should be set or tls pass through enabled, not both")
+	}
+
+	if tlsPassThrough {
+		forwardingRule.TlsPassthrough = tlsPassThrough
```

Comment:
Could we add a comment here describing why we won't need to set the protocol?

That should also remind me of any necessary adjustments once we introduce HTTP/2.